docs/howto/how-to-make-tensorflow-plugin-with-ui/tensorflow-plugin-with-ui.md

```diff
@@ -0,0 +1,1488 @@
+Therefore, we place our `DATA` mode code in its function, `setup_data(self)`.
+Separating your steps into functions naturally helps isolate problem areas in your code, makes debugging much easier, and (like mentioned above) makes even **more** of your code reusable! 
+
+```python
+  def setup_data(self):
+    train_dir = self.provider.tool_config["datasetConfig"]["trainingSetDir"]
+    test_dir = self.provider.tool_config["datasetConfig"]["testSetDir"]
+
+    seed = int(self.provider.tool_config["datasetConfig"]["seed"])
+    batch_size = int(self.provider.tool_config["datasetConfig"]["batchSize"])
+
+    if Path(train_dir).exists():
+      raw_train_ds = tf.keras.utils.text_dataset_from_directory(
+        train_dir, 
+        batch_size=batch_size, 
+        validation_split=0.2, 
+        subset='training', 
+        seed=seed)
+      raw_test_ds = tf.keras.utils.text_dataset_from_directory(
+        test_dir,
+        batch_size=batch_size)
+      raw_val_ds = tf.keras.utils.text_dataset_from_directory(
+        train_dir,
+        batch_size=batch_size, 
+        validation_split=0.2, 
+        subset='validation', 
+        seed=seed)
+      
+      try:
+        tf.data.Dataset.save(raw_train_ds, f"{self.data_url}/train")
+        tf.data.Dataset.save(raw_test_ds, f"{self.data_url}/test")
+        tf.data.Dataset.save(raw_val_ds, f"{self.data_url}/validation")
+        self.info("Completed data prep...")
+      except Exception as e:
+        logger.error(repr(e))
+```
+We know already `self.provider.tool_config` is simply a Python dict.
+Note, though, we have to call `int` on some values. We do this simply because we get all `tool_config` as `string` types from the UI SDK.
+So, we need to convert this to their expected types or even serialize more complex data. (that is enough for its step-by-step guide)
+
+We retrieve and parse these values and use them in `tf.keras.utils.text_dataset_from_directory(...)`. 
+This function loads our raw data into a dataset to be used by TensorFlow.
+Once we've successfully loaded them _as_ a `tf.Dataset`, we `.save(...)` them at distinct locations for our learning setup. Saving them as datasets allows us to use our data in a format optimized for TensorFlow and its training/execution.
+As previously mentioned, we recommend the [companion piece](https://www.tensorflow.org/tutorials/keras/text_classification) provided by TensorFlow for details regarding TensorFlow's API.
+There are more insights regarding TensorFlow specifics, but it's too much to cover here.
+
+Now that our data can be retrieved and loaded, we will need some way to prepare it for our model and training.
+
+### 1.3.2 `PREVIEW` mode
+
+First, we put in our new `PREVIEW` mode code and functionality.
+To better understand TensorFlow usage, please recheck the companion piece for a more detailed explanation of each line.
+_(NOTE: we will assume the reader is doing this or understands `keras` and `tensorflow` moving forward)_
+****
```

Comment:
TW review:

First, we enter the new `PREVIEW` mode code and functionality. To better understand TensorFlow usage, please refer to the companion piece for a more detailed explanation of each line.
****